sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -374,6 +383,29 @@ object ScalaReflection extends ScalaReflection {
           dataType = ObjectType(udt.getClass))
         Invoke(obj, "deserialize", ObjectType(udt.userClass), path :: Nil)
 
+      case t if isValueClass(t) =>
+        val (_, underlyingType) = getUnderlyingParameterOf(t)
+        val underlyingClsName = getClassNameFromType(underlyingType)
+        val clsName = getUnerasedClassNameFromType(t)
+        val newTypePath = s"""- Scala value class: $clsName($underlyingClsName)""" +:
+          walkedTypePath
+
+        // Nested value class is treated as its underlying type
+        // because the compiler will convert value class in the schema to
+        // its underlying type.
+        // However, for value class that is top-level or array element,
+        // if it is used as another type (e.g. as its parent trait or generic),
+        // the compiler keeps the class so we must provide an instance of the
+        // class too. In other cases, the compiler will handle wrapping/unwrapping
+        // for us automatically.
+        val arg = deserializerFor(underlyingType, path, newTypePath)
+        if (walkedTypePath.length == 1 || walkedTypePath.head.contains("array element")) {
```

Comment:
@viirya nice catch! I didn't check for array case.
As you said, basically we just need to instantiate value class if it is the type of array element. However, to do that, in the `case` handling value class, I need to know that I'm checking the type of array element. I think there are 2 ways:
- check the previous piece of `walkedTypePath` like I'm doing here
- introduce a parameter called `previousType: Type` for `deserializeFor`

Personally, I think what I did here (option 1) is a little bit hacky but it is minor and works well. Do you have any suggestion on this?

Reply:
never mind, I think going with option 2 is better in the future